### ISSUE TITLE
cmd/compile: eliminate all (some?) convT2{I,E} calls

### ISSUE THREAD

**Issue by randall77** [golang/go#15375]
For pointer-shaped types, convT2{I,E} are already done by the compiler.  We currently call into the runtime only for non-pointer-shaped types.
There are two cases, where the interface escapes and where it doesn't.
```
type T struct {
    a, b, c int
}
func f(t T) interface{} {
    return t
}
func g(t T) {
	h(t)
}
//go:noescape
func h(interface{})
```
In both cases, I think, it would be easier to inline the work that convT2{I,E} does.  f does this:
```
LEAQ	type."".T(SB), AX
MOVQ	AX, (SP)
LEAQ	"".autotmp_0+40(SP), AX
MOVQ	AX, 8(SP)
MOVQ	$0, 16(SP)
CALL	runtime.convT2E(SB)
MOVQ	24(SP), CX
MOVQ	32(SP), DX
```
instead it could do:
```
LEAQ	type."".T(SB), AX
MOVQ	AX, (SP)
CALL	runtime.newobject(SB)
MOVQ	8(SP), DX
MOVQ	"".autotmp_0+40(SP), AX
MOVQ	"".autotmp_0+48(SP), BX
MOVQ	"".autotmp_0+56(SP), CX
MOVQ	AX, (DX)
MOVQ	BX, 8(DX)
MOVQ	CX, 16(DX)
LEAQ	type."".T(SB), CX
```
11 instructions instead of 8, but several runtime calls bypassed (convT2E, plus the typedmemmove it calls, and everything that calls, ...).
The expanded code gets larger if T has a pointer in it.  Maybe we use typedmemmove instead of explicit copy instructions in that case.

`g` is even easier because there is no runtime call required at all.  Old code:
```
LEAQ	type."".T(SB), AX
MOVQ	AX, (SP)
LEAQ	"".autotmp_2+64(SP), AX
MOVQ	AX, 8(SP)
LEAQ	"".autotmp_3+40(SP), AX
MOVQ	AX, 16(SP)
CALL	runtime.convT2E(SB)
MOVQ	24(SP), CX
MOVQ	32(SP), DX
```
new code:
```
MOVQ	"".autotmp_2+64(SP), AX
MOVQ	"".autotmp_2+72(SP), BX
MOVQ	"".autotmp_2+80(SP), CX
MOVQ	AX, "".autotmp_3+40(SP)
MOVQ	BX, "".autotmp_3+48(SP)
MOVQ	CX, "".autotmp_3+56(SP)
LEAQ	type."".T(SB), CX
LEAQ	"".autotmp_3+40(SP), DX
```
one less instruction, and if we can avoid the copy (if autotmp_2 is never modified after), it could be even better.  And no write barriers are required on the copy even if T has pointers.

Maybe we do the no-escape optimization first.  It seems like an obvious win.  That would allow removing the third arg to convT2{I,E}.  Then we could think about the escape optimization.

@walken-google

**Comment by rsc** [comment]
Sounds like the escape case is for later.

**Comment by griesemer** [comment]
Seems too late now. Moving to 1.11.

**Comment by josharian** [comment]
@randall77 I think this may be done. Do you see anything left to do here?

**Comment by randall77** [comment]
Yes, I think there's more to be done. The escape case (`f` in my example) seems nonoptimal. It makes a temporary, then passes the address of that temporary to `convT2Enoptr`.
We could just call `newobject` directly, then initialize its contents in compiled code. I don't think it would be any bigger (for pointerless types, anyway).
